In [ ]:
# Import libraries
import numpy as np
import matplotlib.pyplot as plt

from ipywidgets import interact, FloatSlider
from IPython.display import display, clear_output

## Reference Table

| Name | Type |
|---|---|
| `numpy` | Package |
| `np` | Alias |
| `matplotlib` | Package |
| `pyplot` | Module |
| `ipywidgets` | Package |
| `interact` | Function |
| `FloatSlider` | Class |
| `IPython` | Package |
| `display` (in `IPython.display`) | Module |
| `display` | Function |
| `clear_output` | Function |

---

## Package, Module, Class, Function — Hierarchy

```
Package
 ├── Module
 │    ├── Class
 │    └── Function
```

---

### 1. Package

A **package** is a directory containing Python code (often with many modules).

- Example: `import numpy`
- The presence of `__init__.py` is a clue that it's a package.

### 2. Module

A **module** is usually a single `.py` file.

```
matplotlib/
│
├── __init__.py
├── pyplot.py
├── axes.py
└── figure.py
```

- Example: `import matplotlib.pyplot`
  - **package**: `matplotlib`
  - **module**: `pyplot`
  - **alias**: `plt`

### 3. Class

A **class** is a blueprint for creating objects.

- Example: `from ipywidgets import FloatSlider`
- You create an instance with `()`:
  ```python
  slider = FloatSlider()
  ```

### 4. Function

A **function** is callable code.

- Example:
  ```python
  from ipywidgets import interact
  from IPython.display import display

  interact(...)
  display(...)
  ```

---

## Practical Clues

- Usually packages/modules are **lowercase**
- Classes are often **PascalCase**
- Functions are often **lowercase**


In [2]:
# Generate synthetic data
np.random.seed(0)
X = np.linspace(0,10,100)
actual_Y = 2 * X + 1 + np.random.normal(0,2,100)

**`np.random.seed(0)`**

Computers don't generate truly random numbers — they use a **pseudo-random number generator (PRNG)**, which produces a sequence that *looks* random. The **seed** is the starting point for that sequence.

- Same seed → same sequence of "random" numbers
- Different seed → different sequence

Why it matters:
- Reproducible experiments
- Others can get the same results
- Easier debugging

---

**`np.linspace(0, 10, 100)`**

`linspace` = **lin**ear **space** → creates evenly spaced numbers between a start and end value.

```
np.linspace(start, stop, num_points)
```

`np.linspace(0, 10, 100)` means:
- Start at `0`
- End at `10`
- Generate `100` evenly spaced values

In [3]:
# Visualization Breakdown
inputs_example = np.linspace(0,10,100) #Numbers from 0-10

np.random.seed(0)
equation_example  = 2 * X + 1 #Equation
random_example = np.random.normal(0,2,100) #Random Noise


#total_example = 2 * X + 1 + np.random.normal(0,2,100) 
total_example = equation_example  + random_example #Total

#First 5 values
print("inputs_example:", inputs_example[:5], "\n")
print("equation_example:", equation_example[:5], "\n")
print("random_example:", random_example[:5], "\n")
print("total_example:", total_example[:5], "\n")

inputs_example: [0.        0.1010101 0.2020202 0.3030303 0.4040404] 

equation_example: [1.         1.2020202  1.4040404  1.60606061 1.80808081] 

random_example: [3.52810469 0.80031442 1.95747597 4.4817864  3.73511598] 

total_example: [4.52810469 2.00233462 3.36151637 6.087847   5.54319679] 



In [4]:
# Create a plot function for interactive visualization
def plot_linear_regression(slope=2.0, intercept=1.0):
    predicted_Y = slope * X + intercept
    loss = np.mean((predicted_Y-actual_Y)**2) # Calculate mean squared error (loss)

    # Create the plot
    plt.figure(figsize=(20,10))
    plt.scatter(X, actual_Y, label='Data Points')
    #plt.plot(X, predicted_Y, color='red', label='Linear Regression Line')
    plt.scatter(X, predicted_Y, color='red', label='Linear Regression Line / Predicted Y')
    plt.xlabel('X')
    plt.ylabel('Actual Y / Predicted Y')
    plt.title(f'Linear Regression (Loss: {loss:.2f})')
    plt.legend()
    plt.grid(True)
    plt.xticks(np.arange(0, 11, 1))   # X grid: 1,2,3,...,10
    plt.yticks(np.arange(0, 45, 2))   #  Y grid: 2,4,6,...,20
    plt.show()

    #First 20 Values
    print("X:", X[:20], "\n")
    print("(Blue) actual_Y:", actual_Y[:20], "\n")
    print("(Red) predicted_Y :", predicted_Y[:20], "\n")


In [5]:
# Create sliders

slope_slider = FloatSlider(value=4.0, min=-5.0, max=5.0, step=0.1, description='Slope')

intercept_slider = FloatSlider(value=3.0, min=-5.0, max=5.0, step=0.1, description='Intercept')

# Create an interactive widget

interact(plot_linear_regression, slope=slope_slider, intercept=intercept_slider)

# Example: Slope = 1.8, Intercept = 2.4

interactive(children=(FloatSlider(value=4.0, description='Slope', max=5.0, min=-5.0), FloatSlider(value=3.0, d…

<function __main__.plot_linear_regression(slope=2.0, intercept=1.0)>

# Why Predicted Values Don't Exactly Match Actual Values

## Question

In real training, shouldn't predicted and actual values be the same if the model is
properly trained? For example, if X = 0 and y_actual = 4.5, shouldn't y_predicted
also be 4.5?

In practice, they're usually *not* exactly the same — and that's expected.

## Answer

### 1. Your data has noise

```python
Y = 2 * X + 1 + np.random.normal(0, 2, 100)
```

The true relationship here is `Y = 2X + 1`, but each data point is intentionally
shifted away from that line by a random noise term. So even a perfect model
of the underlying pattern won't match every individual data point exactly —
because the data points themselves aren't perfectly on the line.

### 2. If the model matched every point perfectly, that's a problem

A model that hits every training point exactly has **overfit** — it memorized
the noise instead of learning the real pattern. This kind of model usually
performs poorly on new, unseen data, because it learned the random fluctuations
as if they were signal.

### 3. An analogy: measuring your height

Imagine measuring your height every day. Due to small measurement errors, you
might get 5'9" one day and 5'10" the next. The "true" underlying model should
estimate something like **5'9.5"** — the best general estimate — not memorize
each individual noisy measurement as if it were the truth.

## Takeaway

> A good model has *some* loss but generalizes well to new data.
> **Zero loss on training data is a red flag, not a success.**

| Scenario | What it means |
|---|---|
| Some loss, good performance on new data | ✅ Model learned the real pattern |
| Zero loss on training data | ⚠️ Likely overfitting (memorized noise) |